mkosi/__init__.py

```diff
@@ -5114,50 +5118,64 @@ def run_verb(args: Args, images: Sequence[Config], *, resources: Path) -> None:
-            die(
-                f"Image '{last.image}' has not been built yet",
-                hint="Make sure to build the image first with 'mkosi build' or use '--force'",
-            )
+    if args.rerun_build_scripts and not output.exists():
+        die(
+            f"Image '{last.image}' must be built once before --rerun-build-scripts can be used",
+            hint="Build the image once with 'mkosi build'",
+        )
 
-        if not last.repart_offline and os.getuid() != 0:
-            die(f"Must be root to build {last.image} image configured with RepartOffline=no")
+    if args.verb != Verb.build and not args.force and not output.exists():
+        die(
+            f"Image '{last.image}' has not been built yet",
+            hint="Make sure to build the image first with 'mkosi build' or use '--force'",
+        )
 
-        check_workspace_directory(last)
+    if not last.repart_offline and os.getuid() != 0:
+        die(f"Must be root to build {last.image} image configured with RepartOffline=no")
 
-        if last.is_incremental():
-            for a, b in itertools.combinations(images, 2):
-                if a.expand_key_specifiers(a.cache_key) == b.expand_key_specifiers(b.cache_key):
-                    die(
-                        f"Image {a.image} and {b.image} have the same cache key '{a.expand_key_specifiers(a.cache_key)}'",  # noqa: E501
-                        hint="Add the &I specifier to the cache key to avoid this issue",
-                    )
+    check_workspace_directory(last)
+
+    if args.rerun_build_scripts and not last.is_incremental():
+        die("Incremental= must be enabled to be able to use --rerun-build-scripts")
 
-        if last.is_incremental() and last.incremental == Incremental.strict:
-            if args.force > 1:
+    if last.is_incremental():
+        for a, b in itertools.combinations(images, 2):
+            if a.expand_key_specifiers(a.cache_key) == b.expand_key_specifiers(b.cache_key):
                 die(
-                    "Cannot remove incremental caches when building with Incremental=strict",
-                    hint="Build once with '-i yes' to update the image cache",
+                    f"Image {a.image} and {b.image} have the same cache key '{a.expand_key_specifiers(a.cache_key)}'",  # noqa: E501
+                    hint="Add the &I specifier to the cache key to avoid this issue",
                 )
 
-            for config in images:
-                if have_cache(config):
-                    continue
+    if last.is_incremental() and (last.incremental == Incremental.strict or args.rerun_build_scripts):
+        if args.force > 1:
+            die(
+                "Cannot remove incremental caches when building with Incremental=strict",
+                hint="Build once with '-i yes' to update the image cache",
+            )
+
+        for config in images:
+            if not config.is_incremental() or have_cache(config):
+                continue
```

Comment:
That both branches below die tripped me for a second before I looked at the other side of the diff.

I think this could be written as
```suggestion
        if any((c := config).is_incremental() and not have_cache(c) for config in images:
```
to reuse the breaking config in the error message. A construction like this was recently used in ukify. That or a small comment above the loop would probably make this easier on the reader.